MCOL-4957 Fix performance slowdown for processing TIMESTAMP columns.

tntnatbry opened the pull request.

tntnatbry committed:
MCOL-4957 Fix performance slowdown for processing TIMESTAMP columns.

As part of MCOL-3776 to address synchronization issue while accessing
the fTimeZone member of the Func class, mutex locks were added to the
accessor and mutator methods. However, this slows down processing
of TIMESTAMP columns in PrimProc significantly as all threads across
all concurrently running queries would serialize on the mutex. This
is because PrimProc only has a single global object for the functor
class (class derived from Func in utils/funcexp/functor.h) for a given
function name. To fix this problem:

  (1) We remove the fTimeZone as a member of the Func derived classes
  (hence removing the mutexes) and instead use the fOperationType
  member of the FunctionColumn class to propagate the timezone values
  down to the individual functor processing functions such as
  FunctionColumn::getStrVal(), FunctionColumn::getIntVal(), etc.

  (2) To achieve (1), a timezone member is added to the
  execplan::CalpontSystemCatalog::ColType class.

tntnatbry committed:
MCOL-4957 Fix performance slowdown for processing TIMESTAMP columns.

Several functors in the Funcexp code call dataconvert::gmtSecToMySQLTime()
and dataconvert::mySQLTimeToGmtSec() functions for conversion between seconds
since unix epoch and broken-down representation. These functions in turn call
the C library function localtime_r() which currently has a known bug of holding
a global lock via a call to __tz_convert. This significantly reduces performance
in multi-threaded applications where multiple threads concurrently call
localtime_r(). More details on the bug:
  https://sourceware.org/bugzilla/show_bug.cgi?id=16145

This bug in localtime_r() caused processing of the Functors in PrimProc to
slowdown significantly since a query execution causes Functors code to be
processed in a multi-threaded manner.

As a fix, we remove the calls to localtime_r() from gmtSecToMySQLTime()
and mySQLTimeToGmtSec() by performing the timezone-to-offset conversion
(done in dataconvert::timeZoneToOffset()) during the execution plan
creation in the plugin. Note that localtime_r() is only called when the
time_zone system variable is set to "SYSTEM".

This fix also required changing the timezone type from a std::string to
a long across the system.

drrtuy commented on dbcon/execplan/treenode.h:
```diff
@@ -1127,6 +1127,17 @@ inline int64_t TreeNode::getDatetimeIntVal()
+        dataconvert::gmtSecToMySQLTime(seconds, m_time, timeZone);
+        dataconvert::DateTime dt(m_time.year, m_time.month, m_time.day, m_time.hour,
+                                 m_time.minute, m_time.second, timestamp.msecond);
+        memcpy(&fResult.intVal, &dt, 8);
```
What do you think about adding static constexpr size_t member to dataconvert::DateTime class and use it in memcpy.
I would better replace memcpy statement with an assignment expression, e.g. fResult.intVal = dt.someMethod()

tntnatbry replied:
`DateTime`, `Time` and `TimeStamp` are 64-bit structs, we cannot add a member variable to them.

drrtuy replied:
Very good. Then Result should have a setter to assign a int value. And this is a very good example why I would like to see setters/getters used instead of direct attributes manipulations. Let me explain. You are setting Result::intVal. Imagine this Result needs to be used as an unsigned int(I am aware about TreeNode used in such context but for the sake of the explanation I will ignore this). Some Result::uIntValue will be used instead. And this potentially gives us hard to detect issues. With a setter you can set a value properly though.

drrtuy commented on dbcon/execplan/functioncolumn.h:
```diff
@@ -219,31 +219,38 @@ class FunctionColumn : public ReturnedColumn
 public:
     virtual const std::string& getStrVal(rowgroup::Row& row, bool& isNull)
     {
+        fOperationType.timeZone = fTimeZone;
```
What do you think if we stick to OOP and use fOperationType::method(fTimeZone) to set the timezeone?

tntnatbry replied:
`fOperationType` here is of type `struct execplan::CalpontSystemCatalog::ColType`. Existing members of this struct are all public and there are no accessor/mutator methods. Changing the struct interface would be outside the scope of this patch.

drrtuy replied:
I am not super crazy about this but I personally think this would allow us to reduce side-effects in the code though. If you don't want to plz resolve this convo. Feel free to introduce setters instead adding a new code. JFYI Bar left them public to reduce his patch.

drrtuy commented on dbcon/mysql/ha_mcs_ddl.cpp:
```diff
@@ -658,11 +658,14 @@ bool anyNullInTheColumn (THD* thd, string& schema, string& table, string& column
-    sf->timeZone(thd->variables.time_zone->get_name()->ptr());
+    const char* timeZone = thd->variables.time_zone->get_name()->ptr();
+    long timeZoneOffset;
+    dataconvert::timeZoneToOffset(timeZone, strlen(timeZone), &timeZoneOffset);
```
We need to discuss this part.
But I suggest to change the API for timeZoneToOffset() and SimpleFilter ctor slightly to reduce side-effects. auto tzOffset = dataconvert::timeZoneToOffset(timeZone, len)); auto* sf = new SimpleFilter(tzOffset, sop);

tntnatbry replied:
`timeZoneToOffset` returns a bool value indicating whether the conversion passed or failed. This return value is used in cpimport while parsing command line arguments, that's why we pass the offset by reference as a function argument. Changing `SimpleFilter` API would involve changing code at other places as well where we call this ctor, this would be outside the scope of this patch.

drrtuy replied:
Agree.

drrtuy commented on dbcon/mysql/ha_mcs_ddl.cpp:
```diff
@@ -1127,7 +1134,7 @@ int ProcessDDLStatement(string& ddlStatement, string& schema, const string& tabl
             }
 
-            alterTable->fTimeZone.assign(thd->variables.time_zone->get_name()->ptr());
+            alterTable->fTimeZone = timeZoneOffset;
```
It is worth to discuss to use method to set alterTable::fTimeZone

tntnatbry replied:
All existing members of `struct AlterTableStatement` are public.

drrtuy replied:
Plz see my prev comments regarding direct assignments to public attributes.

tntnatbry commented:
@drrtuy I have created a new PR for develop-6 to rebase over the clang-format commit, that addresses the comments above: https://github.com/mariadb-corporation/mariadb-columnstore-engine/pull/2262. So I am closing this PR.
